### PR TITLE
Add unit test for torch_geometric library

### LINKED ISSUE #65138: GNN (SAGE Convolutional) ONNX export
## 🐛 Bug

Hello Everyone, So past few days I have been trying to run GNN using ONNX runtime (Python and eventually on C++). However I get the following error after I successfully export the onnx model from pytorch which uses pytorch geometric layers (SAGE Convolution to be precise) and functions for the GNNs. Before blaming the ONNX runtime I want to make sure that the ONNX exporter indeed FULLY supports SAGEConv conversion. When I try to run the converted ONNX model I get the following error: 

`InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Non-zero status code returned while running ScatterElements node. Name:'ScatterElements_26' Status Message: Indices dim=12 at pos=0 is greater than input dim=6`
## To Reproduce

Here is the code I used for pytorch model creation:
```python
#%debug
import torch
import torchvision
import torch_geometric
import torch.nn.functional as F
from torch.nn import Linear as Lin, Sequential as Seq, ReLU, AvgPool1d, Dropout, Softmax, Sigmoid, BatchNorm1d
from torch_geometric.nn import EdgeConv, DynamicEdgeConv, BatchNorm, avg_pool_x, knn_graph, EdgePooling, DenseSAGEConv, dense_diff_pool, dense_mincut_pool, SAGEConv, max_pool_x, SAGPooling, JumpingKnowledge, global_mean_pool as gap, global_max_pool as gmp
#from torch_cluster import knn_graph

#dummy_input = (torch.randn(10,5, device='cuda'),torch.randn(2,10, device='cuda').type(torch.int64))
coords0 = torch.randn(1,6)
coords1 = torch.randn(1,6)


coords = torch.transpose(torch.cat((coords0,coords1),dim=0),0,1)
adj = knn_graph(coords, k=2, batch=None, loop=True)

edge_from = adj[0:1,:]
edge_to = adj[1:,:]

input_values = (coords0,coords1, edge_from, edge_to)
#model = torchvision.models.alexnet(pretrained=True).cuda()
class trial1(torch.nn.Module):
    def __init__(self):
        super(trial1, self).__init__()
        
        self.SAGEConvBlock1 = SAGEConv(2, 512, normalize=True)
        self.bano1 = BatchNorm(512)
        self.relu = ReLU()
        self.dense1 = Seq(Lin(512,1))
        self.sigmoid = Sigmoid()
    def forward(self, coords0, coords1, edge_from, edge_to):
      adj = torch.cat((edge_from,edge_to),dim=0)
      gra = torch.transpose(torch.cat((coords0,coords1),dim=0),0,1)
      
      x1 = self.SAGEConvBlock1(gra, edge_index=adj)
      x = torch.unsqueeze(torch.sum(x1), dim=0)
      return x

model2 = trial1()

input_names = ["coords0","coords1", "edge_from", "edge_to"]
output_names = [ "outputs" ]

torch.onnx.export(model2, input_values, "ONNX_models/trial2.onnx", verbose=True, input_names=input_names, output_names=output_names, opset_version=11, dynamic_axes={'coords0':{0:'batch_size',1:'features'}, 'coords1':{0:'batch_size',1:'features'}, 'edge_from':{0:'batch_size',1:'features'}, 'edge_to':{0:'batch_size',1:'features'}, 'outputs':{0:'batch_size'}})
```

The onnx runtime code to perform the inference is:
```python
import onnxruntime
print(onnxruntime.__version__)
ort_session = onnxruntime.InferenceSession("ONNX_models/trial2.onnx")

def to_numpy(tensor):
    return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()

ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(coords0), ort_session.get_inputs()[1].name: to_numpy(coords1), ort_session.get_inputs()[2].name: to_numpy(edge_from), ort_session.get_inputs()[3].name: to_numpy(edge_to)}
ort_outs = ort_session.run(["outputs"], ort_inputs)
``` 

I get the error in the last line of the above code: 
`ort_outs = ort_session.run(["outputs"], ort_inputs)`

 - PyTorch Version (e.g., 1.0): 1.9.0
 - OS (e.g., Linux): Linux/Google Colab
 - How you installed PyTorch (`conda`, `pip`, source): pip
 
 - Python version: 3.x
 - CUDA/cuDNN version: 10.2
 - GPU models and configuration: Google Colab (K80)
 - Any other relevant information:

## Additional context

Also, there is no error and the inference runs perfectly fine when I remove the SAGEConv layer from the model. The technical issue is the dimension mismatch between gra (shape: [6,2] )  and adj (shape: [2,36] )  where apparently the gra.shape[0] should be equal to adj.shape[1]. And it indeed works if we use only the first 6 elements of the adj (shape: [2,6] ) which essentially is not what we would intend to do. For my application I need to have multiple inputs of shape ('batch_size',nodes) and edges of the shape ('batch_size', k*nodes) where k indicates the number of k nearest neighbours.

### Notes
I would really appreciate if someone has any leads on this or if anyone has successfully deployed a GNN on ONNX using different input dimensions (where rank of each input <= 2).

After verification, I am inclined towards believing that the issue is with the ONNX exporter which when creating ONNX compatible graph disturbs the shapes of the tensors. ONNX exporter is not yet mature enough to handle Pytorch Geometric's GNN functions like SAGEConv, etc. This needs to be improved.

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Fixes #65138

### CODE REVIEW

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 1]
```diff
@@ -36,6 +36,7 @@
     skipTraceTest,
 )
 from torch import Tensor
+from torch.hub import _check_module_exists
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:39]: Avoid importing functions. Only import modules. https://github.com/pytorch/pytorch/wiki/torch.onnx-Namespacing

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 2]
```diff
@@ -12480,6 +12481,58 @@ def forward(self, x):
...
+            Linear as Lin,
+            Sequential as Seq,
+            ReLU,
+            Sigmoid,
+        )
+        from torch_geometric.nn import (
+            BatchNorm,
+            knn_graph,
+            SAGEConv,
+        )
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:12498]: Avoid importing functions. Only import modules. https://github.com/pytorch/pytorch/wiki/torch.onnx-Namespacing

**Review by justinchuby — CHANGES_REQUESTED** [review]

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -12480,6 +12480,60 @@ def forward(self, x):
...
         self.run_test(LerpModel(), torch.rand(5, 4, 3))
 
+    @skipScriptTest()
+    @skipIfUnsupportedMinOpsetVersion(16)
+    @unittest.skipIf(
+        not hub._check_module_exists("torch_geometric"),
+        "torch_geometric not installed.",
+    )
+    def test_sage_conv(self):
+        from torch import nn as torch_nn
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:12490]: I would move this import to the top. torch.nn is always available.

**Review by justinchuby — APPROVED** [review]

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 4]
```diff
@@ -35,7 +35,7 @@
     skipShapeChecking,
     skipTraceTest,
 )
-from torch import Tensor
+from torch import hub, Tensor
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:38]: Since hub is use only once, I would just call `torch.hub._check_module_exists` where it is used to keep the code clean. `torch.hub` is not too long anyways.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 5]
```diff
@@ -12480,6 +12480,60 @@ def forward(self, x):
 
         self.run_test(LerpModel(), torch.rand(5, 4, 3))
 
+    @skipScriptTest()
+    @skipIfUnsupportedMinOpsetVersion(16)
+    @unittest.skipIf(
+        not hub._check_module_exists("torch_geometric"),
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:12486]:
```suggestion
        not torch.hub._check_module_exists("torch_geometric"),
```

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 4 above)
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:38]:
```suggestion
from torch import Tensor
```

Since hub is use only once, I would just call `torch.hub._check_module_exists` where it is used to keep the code clean. `torch.hub` is not too long anyways.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 6]
```diff
@@ -12480,6 +12480,60 @@ def forward(self, x):
...
+        inputs = (coords0, coords1, edge_from, edge_to)
+
+        class MySAGEConv(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.SAGEConvBlock1 = torch_geometric_nn.SAGEConv(
+                    2, 512, normalize=True
+                )
+                self.bano1 = torch_geometric_nn.BatchNorm(512)
+                self.relu = torch_nn.ReLU()
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:12509]:
```suggestion
                self.relu = torch.nn.ReLU()
```

here and below. As `torch_nn` doesn't save us space

File: test/onnx/test_pytorch_onnx_onnxruntime.py (hunk 2 above)
Comment (thiagocrepaldi) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: This is narrowed down to a function namespace. It doesn't pollute API neither cause circular imports

Comment (justinchuby) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: That's great, still: unqualified names are harder to read because it's harder to know where they come from. If we always import modules, we can assume that the unqualified ones are defined within the same module. Also importing inside a function is discouraged. When possible always import on the top, unless there are dependency/ performance concerns.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 7]
```diff
@@ -12480,6 +12480,59 @@ def forward(self, x):
 
         self.run_test(LerpModel(), torch.rand(5, 4, 3))
 
+    @skipScriptTest()
+    @skipIfUnsupportedMinOpsetVersion(16)
+    @unittest.skipIf(
+        not torch.hub._check_module_exists("torch_geometric"),
```
Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:12486]: Do we need to install torch_geometric in our test env?

Comment (justinchuby) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:12486]: _check_module_exists is a private method. I suggest duplicating it here

```
def _check_module_exists(name):
    import importlib.util
    return importlib.util.find_spec(name) is not None
```

**Comment by BowenBao** [comment]
> Fixes #65138

How does adding a unit test fix the issue?

**Comment by justinchuby** [comment]
Looking at this again - I think we should treat the CI as the source of truth and have all tests run in it. Having a test that doesn't run in CI is a little strange.

**Comment by thiagocrepaldi** [comment]
> > Fixes #65138
> 
> How does adding a unit test fix the issue?

It doesn't. The term "Fixes #Issue" is used just to close the issue when the PR is merged. I didn't want to close the issue without official "proof" and this PR serves as one

**Review by BowenBao — APPROVED** [review]
